File: C/24_DFS&BFS/BOJ_11_1707_이분그래프.c
#include <stdio.h>
#include <stdlib.h>
int v, e, tst, ** arr, * visited, * checked, * cnt;
void dfs(int idx, int side) {
	if (!tst) return;
	visited[idx] = 1;
	checked[idx] = side;
	for (int i = 0; i < cnt[idx]; i++) {
		int tmp = arr[idx][i];
		if (!visited[tmp]) {
			dfs(tmp, !side);
		}
		else if (checked[tmp] == side) {
			tst = 0;
			return;
		}
	}
}
main() {
	int k, idx = 0;
	scanf("%d", &k);
	int* result = (int*)malloc(sizeof(int) * k);
	while (k--) {
		int* xi, * yi;
		scanf("%d %d", &v, &e);
		arr = (int**)malloc(sizeof(int*) * (v + 1));
		cnt = (int*)malloc(sizeof(int) * (v + 1));
		visited = (int*)malloc(sizeof(int) * (v + 1));
		checked = (int*)malloc(sizeof(int) * (v + 1));
		xi = (int*)malloc(sizeof(int) * (e + 1));
		yi = (int*)malloc(sizeof(int) * (e + 1));
		for (int i = 1; i <= e; i++) {
			int x, y;
			scanf("%d %d", &x, &y);
			cnt[x]++;
			cnt[y]++;
			xi[i] = x;
			yi[i] = y;
		}
		for (int i = 1; i <= v; i++) {
			arr[i] = (int*)malloc(sizeof(int) * cnt[i]);
			visited[i] = 0;
			cnt[i] = 0;
		}
		for (int i = 1; i <= e; i++) {
			arr[xi[i]][cnt[xi[i]]++] = yi[i];
			arr[yi[i]][cnt[yi[i]]++] = xi[i];
		}
		tst = 1;
		for (int i = 1; i <= v; i++) {
			if (!visited[i]) {
				dfs(i, 1);
			}
		}
		if (tst) {
			result[idx++] = 1;
		}
		else {
			result[idx++] = 0;
		}
	}
	for (int i = 0; i < idx; i++) {
		if (result[i]) {
			printf("YES\n");
		}
		else {
			printf("NO\n");
		}
	}
}

File: JAVA/24_DFS&BFS/BOJ_11_1707_이분그래프.java
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Queue;
import java.util.LinkedList;
import java.util.StringTokenizer;

public class Main {

    static int v, e;
    static boolean tst;
    static boolean[] visited, checked;
    static ArrayList<Integer>[] list;
    static Queue<Integer> queue;
    static StringBuilder sb;

    public static void bfs(int idx) {
        queue.offer(idx);
        checked[idx] = true;
        while (!queue.isEmpty()) {
            int tmp = queue.poll();
            visited[tmp] = true;
            if (list[tmp] != null) {
                for (int i : list[tmp]) {
                    if (!visited[i]) {
                        checked[i] = !checked[tmp];
                        queue.offer(i);
                    } else if (checked[tmp] == checked[i]) {
                        sb.append("NO\n");
                        tst = false;
                        return;
                    }
                }
            }
        }
    }

    @SuppressWarnings("unchecked")
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int k = Integer.parseInt(br.readLine());
        sb = new StringBuilder();
        while (k-- > 0) {
            StringTokenizer st = new StringTokenizer(br.readLine(), " ");
            v = Integer.parseInt(st.nextToken());
            e = Integer.parseInt(st.nextToken());
            visited = new boolean[v + 1];
            checked = new boolean[v + 1];
            list = new ArrayList[v + 1];
            for (int i = 1; i <= v; i++) {
                list[i] = new ArrayList<Integer>();
            }
            for (int i = 0; i < e; i++) {
                st = new StringTokenizer(br.readLine(), " ");
                int x = Integer.parseInt(st.nextToken());
                int y = Integer.parseInt(st.nextToken());
                list[x].add(y);
                list[y].add(x);
            }
            tst = true;
            queue = new LinkedList<>();
            for (int i = 1; i <= v; i++) {
                if (!tst) break;
                if (!visited[i]) {
                    bfs(i);
                }
            }
            if (tst) sb.append("YES\n");
        }
        System.out.print(sb);
    }
}